src/twisted/internet/defer.py

```diff
@@ -1459,8 +1459,11 @@ def _inlineCallbacks(result, g, status):
             # code.
             appCodeTrace = exc_info()[2].tb_next
 
-            # The contextvars backport and our no-op shim add an extra frame.
             if version_info < (3, 7):
+                # The contextvars backport and our no-op shim add an extra frame.
+                appCodeTrace = appCodeTrace.tb_next
+            elif implementation.name == "pypy":
```

Comment:
Looks like pypy 3.5, pypy 3.6, and pypy 3.7 behave differently.
Let's have this logic work on all three.  Then after this release we can drop pypy 3.5

Reply:
Current PyPy doesn't support 3.5.  Do we need to add it into the mix?  I'm not sure offhand if the action provides it but we can try.

Reply:
Looks like https://github.com/actions/setup-python#available-versions-of-pypy
specifies the versions of Pypy.  I would say going through hoops to get Pypy 3.5  in Twisted CI is not worth the effort at this point.

Since we we specified `pypy3` before, and that was installing pypy3 compatible with Python 3.6.12, then I see no point in introducing Pypy 3.5 at this point in the release cycle, when we are moving to drop support for it after the release.

Reply:
Can you manually install the last Pypy 3.5 release from:
https://downloads.python.org/pypy/pypy3.5-v7.0.0-linux64.tar.bz2
https://downloads.python.org/pypy/pypy3.5-v7.0.0-osx64.tar.bz2

And manually run `test_inlinecb.py` in your environment with your patch with this version of pypy?

I just want to double check, just in case there is a remote possibility that someone out there is running Pypy 3.5

Reply:
```console
$ git rev-parse HEAD
a225560007f04235596992da561adcf686c118ab
```

```console
$ git diff
diff --git a/src/twisted/logger/_buffer.py b/src/twisted/logger/_buffer.py
index f90f8f982..3334f97e9 100644
--- a/src/twisted/logger/_buffer.py
+++ b/src/twisted/logger/_buffer.py
@@ -7,7 +7,7 @@ Log observer that maintains a buffer.
 """
 
 from collections import deque
-from typing import Deque, Optional
+from typing import Optional
 
 from zope.interface import implementer
 
```

```console
$ venvpp35/bin/python --version --version
Python 3.5.3 (928a4f70d3de7d17449456946154c5da6e600162, Feb 09 2019, 11:50:43)
[PyPy 7.0.0 with GCC 8.2.0]
```

```console
$ venvpp35/bin/trial twisted.internet.test.test_inlinecb
twisted.internet.test.test_inlinecb
  CancellationTests
    test_asynchronousCancellation ...                                      [OK]
    test_cascadeCancellingOnCancel ...                                     [OK]
    test_errbackCancelledErrorOnCancel ...                                 [OK]
    test_errorToErrorTranslation ...                                       [OK]
    test_errorToSuccessTranslation ...                                     [OK]
  ForwardTraceBackTests
    test_forwardLotsOfTracebacks ...                                       [OK]
    test_forwardTracebacks ...                                             [OK]
  NonLocalExitTests
    test_returnValueNonLocalDeferred ...                                   [OK]
    test_returnValueNonLocalWarning ...                                    [OK]
  StopIterationReturnTests
    test_returnWithValue ...                                               [OK]

-------------------------------------------------------------------------------
Ran 10 tests in 0.135s

PASSED (successes=10)
```

Reply:
Also full tox passed.

https://gist.github.com/altendky/8fab0cac9a0be55f2d836bcc018c6637

Reply:
That's good enough for me.
No need change the CI to add Pypy 3.5.  If someone encounters a problem, we can point them to this
log, and ask them to upgrade to Pypy 3.6 which is better tested/supported.